From bb69c346238fa0f9b340995e0965221503a4858a Mon Sep 17 00:00:00 2001
From: gsheffi <gsheffi@yahoo-inc.com>
Date: Thu, 20 Sep 2018 11:03:28 +0300
Subject: [PATCH] Before rebasing

---
 .../benchmark/indexing/OakIncrementalIndexReadBenchmark.java    | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/benchmarks/src/main/java/org/apache/druid/benchmark/indexing/OakIncrementalIndexReadBenchmark.java b/benchmarks/src/main/java/org/apache/druid/benchmark/indexing/OakIncrementalIndexReadBenchmark.java
index dfedd4151..02bdacfe7 100644
--- a/benchmarks/src/main/java/org/apache/druid/benchmark/indexing/OakIncrementalIndexReadBenchmark.java
+++ b/benchmarks/src/main/java/org/apache/druid/benchmark/indexing/OakIncrementalIndexReadBenchmark.java
@@ -74,7 +74,7 @@ import java.util.concurrent.TimeUnit;
 @Measurement(iterations = 25)
 public class OakIncrementalIndexReadBenchmark
 {
-  @Param({"75000"})
+  @Param({"150000"})
   private int rowsPerSegment;
 
   @Param({"basic"})
-- 
2.17.1 (Apple Git-112)

